sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,731 @@
+            typename = std::enable_if_t<__T::is1D()>>
+  auto select(uint16_t __Offset = 0) &&;
+  template <int __SizeY, int __StrideY, int __SizeX, int __StrideX,
+            typename __T = type, typename = std::enable_if_t<__T::is2D()>>
+  auto select(uint16_t __OffsetY = 0, uint16_t __OffsetX = 0) &;
+  template <int __SizeY, int __StrideY, int __SizeX, int __StrideX,
+            typename __T = type, typename = std::enable_if_t<__T::is2D()>>
+  auto select(uint16_t __OffsetY = 0, uint16_t __OffsetX = 0) &&;
+
+  // Replicate operations.
+  template <int Rep> simd<element_type, Rep> replicate();
+  template <int Rep, int W>
+  simd<element_type, Rep * W> replicate(uint16_t OffsetX);
+  template <int Rep, int W>
+  simd<element_type, Rep * W> replicate(uint16_t OffsetY,
+                                        uint16_t OffsetX);
+  template <int Rep, int VS, int W>
+  simd<element_type, Rep * W> replicate(uint16_t OffsetX);
+  template <int Rep, int VS, int W>
+  simd<element_type, Rep * W> replicate(uint16_t OffsetY,
+                                        uint16_t OffsetX);
+  template <int Rep, int VS, int W, int HS>
+  simd<element_type, Rep * W> replicate(uint16_t OffsetX);
+  template <int Rep, int VS, int W, int HS>
+  simd<element_type, Rep * W> replicate(uint16_t OffsetY,
+                                        uint16_t OffsetX);
+
+  // Format operations.
+  template <typename __EltTy> auto format();
+  template <typename __EltTy> auto format() &&;
+  template <typename __EltTy, int __Height, int __Width> auto format() &;
+  template <typename __EltTy, int __Height, int __Width> auto format() &&;
+
+  // Merge operations.
+  void merge(const value_type &__Val, const __mask_type_t<length> &__Mask);
+  void merge(const value_type &__Val1, value_type __Val2,
+             const __mask_type_t<length> &__Mask);
+} // namespace gen
+} // namespace cm
+```
+
+```sycl::intel::gpu::simd_view``` class supports all the element-wise operations and
+other utility functions defined for ```sycl::intel::gpu::simd``` class. It also
+provides region accessors and more generic operations tailored for 2D regions,
+such as row/column operators and 2D select/replicate/format/merge operations.
+
+```cpp
+  simd<float, 32> v1;
+  auto m1 = v1.format<float, 4, 8>();
+  simd<float, 4> v2;
+  auto m2 = v2.format<float, 2, 2>();
+
+  // ...
+  m2 = m1.select<2, 2, 2, 4>(1, 2);  // v_size = 2, v_stride = 2,
+                                     // h_size = 2, h_stride = 4,
+                                     // v_offset = 1, h_offset = 2.
+```
+<p align="center">
+<img src="images/Matrix_2_2_2_4__1_2.svg" title="2D select example" width="700" height="200"/>
+</p>
+
+```cpp
+  m1.select<4, 1, 4, 2>(0, 0) = 0.0f; // selected elements of m1
+                                      // are replaced with 0.0f
+```
+
+<p align="center">
+<img src="images/Matrix_4_1_4_2__0_0.svg" title="2D select example" width="700" height="200"/>
+</p>
+
+### Reduction functions
```

Comment:
We need to rename all the functions in this section (because they start with `cm_`).  I think we should use this as an opportunity to align with the reductions in `std::simd` and our other proposals.

Based on `std::simd` in [Parallelism TS v2](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/n4808.pdf), we could use:
```
template <typename T1, int SZ, typename BinaryOperation>
reduce(simd<T1, SZ> v, BinaryOperation op);

reduce(simd<T1, SZ> v, std::plus<>()) // equivalent to cm_sum
reduce(simd<T1, SZ> v, std::multiplies<>()) // equivalent to cm_prod
reduce(simd<T1, SZ> v, sycl::maximum<>()) // equivalent to cm_reduced_max
reduce(simd<T1, SZ> v, sycl::minimum<>()) // equivalent to cm_reduced_min
```

We should limit support to these four built-in reduction operators initially, but this gives us a nice extension point for things like user-defined reduction operators.

If we think that having yet another overload of `reduce()` is unclear, we could call these something like `hreduce()` (horizontal reduce) or `simd_reduce()`.

Reply:
@kychendev , please address

Reply:
The  comments make sense to me.

Reply:
These comments haven't been addressed yet.  If @kychendev isn't opposed to the direction, I think we should rename these functions before the merge.  It seems like a small change to me.

Reply:
I have updated the following reduction APIs as suggested:
reduce(simd<T1, SZ> v, std::plus<>())
reduce(simd<T1, SZ> v, std::multiplies<>())
For max/min I changed to the following as that's how it's defined in Parallelism TS v2, 
hmax(simd<T1, SZ> v)
hmin(simd<T1, SZ> v)